File: plugin/dropbox/www/dropbox.js
window.plugin = {
  name: 'Dropbox',
  loaded: 'dropbox_plugin_loaded',
  onselect: 'dropbox_plugin_update'
};


function dropbox_plugin_loaded() {

    $('#DropboxGetAuthKeyButton').click(function(){
	    $.ajax({
	        type: "GET",
	        async: false, // avoid popup blocker
	        url: "/plugin/dropbox/dropbox.py/authurl",
	        contentType: "application/json; charset=utf-8",
	        success: function(response){
				if (response.Result == 'ERROR')
					alert(response.Message);
				else
	    			window.open(response.Url);
	        },
	    });
		return false;
    });

    $('#DropboxSubmitAuthKeyButton').click(function(){
	  	var dataString = 'code=' + $('#DropboxAuthCode').val();
		$.ajax({
			type: "POST",
			url: "/plugin/dropbox/dropbox.py/authorize",
			data: dataString,
			dataType: "json",
			success: function(response) {
				if (response.Result == 'OK') {
					//alert('Authorization done.');
					$('#DropboxAuthCode').val('');
					dropbox_plugin_update();
				}
				else
					alert(response.Message);
			},
			error: function (xhr, ajaxOptions, thrownError) {
	    		alert('Error: ' + xhr.status + ' - ' + thrownError);
	  		},
		});
		return false;
    });

    $('#DropboxDisconnectButton').click(function(){
    	if (confirm('Do you really want to disconnect?')) {
		    $.ajax({
		        type: "GET",
		        url: "/plugin/dropbox/dropbox.py/disconnect",
		        contentType: "application/json; charset=utf-8",
		        success: function(response){
		        	dropbox_plugin_update();
		        },
				error: function (xhr, ajaxOptions, thrownError) {
		    		alert('Error: ' + xhr.status + ' - ' + thrownError);
		  		},
		    });
		}
    });

    $('#DropboxDeleteFilesButton').click(function(){
    	if (confirm('Do you really want to delete all images?')) {
    		$("body").css("cursor", "progress");
		    $.ajax({
		        type: "GET",
		        url: "/plugin/dropbox/dropbox.py/delete_files",
		        contentType: "application/json; charset=utf-8",
		        success: function(response){
		        	dropbox_plugin_update();
					$("body").css("cursor", "default");
		        },
				error: function (xhr, ajaxOptions, thrownError) {
					$("body").css("cursor", "default");
		    		alert('Error: ' + xhr.status + ' - ' + thrownError);
		  		},
		    });
		}
    });
    
	$('#DropboxSaveButton').click(function(){

		var dataString = 'active=' + $("#DropboxUpload").prop('checked');
		$.ajax({
			type: "POST",
			url: "/plugin/dropbox/dropbox.py/set_config_data",
			data: dataString,
			dataType: "json",
			success: function(response) {
				if (response.Result == 'ERROR')
					alert(response.Message);
			},
			error: function (xhr, ajaxOptions, thrownError) {
				alert('Error: ' + xhr.status + ' - ' + thrownError);
			},
		});
		return false;
	});

	$('#DropboxCancelButton').click(function(){
		dropbox_plugin_update();
		return false;
	});

    dropbox_plugin_update();
}


function dropbox_plugin_update() {
    $.ajax({
        type: "GET",
        url: "/plugin/dropbox/dropbox.py/status",
        contentType: "application/json; charset=utf-8",
        success: function(response){
			$('#DropboxDisplayName').empty()
			$('#DropboxDisplayName').append(response.Info);
			if (response.Result == 'ERROR') {
				alert(response.Message);
			}
			else {
				if (response.IsConnected) {
					$('#DropboxFiles').empty()
					$('#DropboxFiles').append('Files: ' + response.Files);
					$('#DropboxRegContainer').hide();
					$('#DropboxDetails').show();
					$('#DropboxUnregContainer').show();
					if (response.Files == 0) {
						$('#DropboxDeleteFilesButton').prop("disabled", true);
					}
					else {
						$('#DropboxDeleteFilesButton').prop("disabled", false);
					}
				}
				else {
					$('#DropboxRegContainer').show();
					$('#DropboxDetails').hide();
					$('#DropboxUnregContainer').hide();
				}
			}
        }
    });
    
    $.ajax({
        type: "GET",
        url: "/plugin/dropbox/dropbox.py/get_upload_captures",
        contentType: "application/json; charset=utf-8",
        success: function(response){
			if (response.Result == 'ERROR') {
				alert(response.Message);
			}
			else {
	        	$("#DropboxUpload").prop('checked', (response.Active == true));
			}
        }
    });
}

File: plugin/dropbox/www/dropbox.py
# Watchdog - Building supervision with Raspberry Pi
# (C)2016  - Norbert Huffschmid - GNU GPL V3 

import ConfigParser
import dbus
import json
import time
import watchdog.config

CONFIG_FILE = '/etc/watchdog.conf'


def status(req):
    req.content_type = 'application/json; charset=UTF8'
    
    trialCount = 5
    errorMessage = ''
    while (trialCount > 0):
    
        try:
            bus = dbus.SystemBus()
            proxy = bus.get_object('net.longexposure.watchdog.plugin.dropbox', '/')
            broker = dbus.Interface(proxy, 'net.longexposure.watchdog.plugin.dropbox')
            status = broker.get_status()
            return json.dumps({ 'Result': 'OK', 'IsConnected': status['IsConnected'], 'Info': status['Info'], 'Files': status['Files'] })
    
        except dbus.DBusException, e:
            # plugin startup takes some time ==> make some trials before terminating with error
            time.sleep(5)
            trialCount = trialCount - 1
            errorMessage = str(e)

        except Exception, e:
            return json.dumps({ 'Result': 'ERROR', 'Message': 'Getting status failed: ' + str(e) })

    return json.dumps({ 'Result': 'ERROR', 'Message': 'Getting status failed after 5 trials: ' + errorMessage})


def authurl(req):
    req.content_type = 'application/json; charset=UTF8'

    try:
        
        bus = dbus.SystemBus()
        proxy = bus.get_object('net.longexposure.watchdog.plugin.dropbox', '/')
        broker = dbus.Interface(proxy, 'net.longexposure.watchdog.plugin.dropbox')
        url = broker.get_authorization_url()
        return json.dumps({ 'Result': 'OK', 'Url': url })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'authurl failed: ' + str(e) })


def authorize(req):
    req.content_type = 'application/json; charset=UTF8'

    try:
        
        # read POST parameter
        code = req.form.getfirst('code')

        bus = dbus.SystemBus()
        proxy = bus.get_object('net.longexposure.watchdog.plugin.dropbox', '/')
        broker = dbus.Interface(proxy, 'net.longexposure.watchdog.plugin.dropbox')
        success = broker.authorize(code)

        if (success):
            return json.dumps({ 'Result': 'OK' })
        else:
            return json.dumps({ 'Result': 'ERROR', 'Message': 'Authorization failed' })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'authorize failed: ' + str(e) })


def disconnect(req):
    req.content_type = 'application/json; charset=UTF8'

    try:

        bus = dbus.SystemBus()
        proxy = bus.get_object('net.longexposure.watchdog.plugin.dropbox', '/')
        broker = dbus.Interface(proxy, 'net.longexposure.watchdog.plugin.dropbox')
        broker.disconnect()
        return json.dumps({ 'Result': 'OK' })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'disconnect failed: ' + str(e) })


def delete_files(req):
    req.content_type = 'application/json; charset=UTF8'

    try:

        bus = dbus.SystemBus()
        proxy = bus.get_object('net.longexposure.watchdog.plugin.dropbox', '/')
        broker = dbus.Interface(proxy, 'net.longexposure.watchdog.plugin.dropbox')
        broker.delete_files()
        return json.dumps({ 'Result': 'OK' })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'delete_files failed: ' + str(e) })
    
    
def get_upload_captures(req):
    req.content_type = 'application/json; charset=UTF8'

    try:
        return json.dumps({ 'Result': 'OK', 'Active': (watchdog.config.get_option('dropbox', 'active') == 'true') })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'get_upload_captures failed: ' + str(e) })


def set_config_data(req):
    req.content_type = 'application/json; charset=UTF8'

    try:
        watchdog.config.write_option('dropbox', 'active', req.form.getfirst('active'))
        return json.dumps({ 'Result': 'OK' })

    except Exception, e:
        return json.dumps({ 'Result': 'ERROR', 'Message': 'Cannot set config data: ' + str(e) })
